Add BaseDiscreteElement

tteschemacher opened the pull request:
The element is the new basis class for all discrete IGA elements.

tteschemacher committed:
Add BaseDiscreteElement

The element is new the basis class for all discrete IGA elements.

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h:
```diff
@@ -0,0 +1,449 @@
+    * @param[in] rWorkingSpaceDimension GeometrySpace coordinates. For surfaces default 3.
+    * @param[in] rLocalSpaceDimension ParameterSpace coordinates For surfaces default 2.
+    */
+    void Jacobian(const Matrix& DN_De,
```
didn't you say this should be virtual?

loumalouomega replied:
const int& == const int

tteschemacher replied:
The function is valid for any geometry object due to the space dimension inputs.

tteschemacher replied:
I changed the int& to int.

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h:
```diff
@@ -0,0 +1,449 @@
+    * @param rRightHandSideVector the elemental right hand side vector
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void CalculateRightHandSide(
```
I think you are missing ClacLHS

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h:
```diff
@@ -0,0 +1,449 @@
+
+
+/********************************************************************/
+/*    CalculateOnIntegrationPoints                                  */
```
Why do you override all these functions even though they do nothing?
=> seems like a duplication of `element.h`
Plus keep in mind that this changes in #2194 !

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h:
```diff
@@ -0,0 +1,449 @@
+    * @param CalculateStiffnessMatrixFlag: The flag to set if compute the LHS
+    * @param CalculateResidualVectorFlag: The flag to set if compute the RHS
+    */
+    virtual void CalculateAll(
```
If you do it this way then please implement it efficient, i.e. if only the LHS or RHS is required that the other one is not computed

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.h:
```diff
@@ -0,0 +1,449 @@
+#if !defined(KRATOS_BASE_DISCRETE_ELEMENT_H_INCLUDED )
```
Missing header

philbucher reviewed (commented):
Some comments, mostly regarding the GP-stuff

philbucher commented:
also take a look at the codacy suggestions, they are relevant

loumalouomega reviewed (approved):
Take into account the comments from @philbucher. I agree with him

loumalouomega reviewed (changes requested):
Minor commentaries

tteschemacher committed:
Fixes in:
SetValueOnIntegrationPoints -> SetValuesOnIntegrationPoints
CalculateOnIntegrationPoints -> Calculate
Header added
CalculateLeftHandSide added

tteschemacher committed:
const int& -> const int

tteschemacher committed:
Bugfix in GetValuesVector
i -> index

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+                    GetValue(SHAPE_FUNCTION_VALUES)
+                );
+            }
+            else
```
brackets

tteschemacher replied:
I would say it makes it much more readable

oberbichler replied:
Just an advice... I don't think that the readability changes but leaving them off can lead to semantic errors in the code.

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+    {
+        KRATOS_TRY
+
+            if (GetProperties()[CONSTITUTIVE_LAW] != nullptr) {
```
no indention

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        KRATOS_TRY
+
+        if (GetProperties()[CONSTITUTIVE_LAW] != nullptr) {
+                mConstitutiveLawVector[0]->ResetMaterial(GetProperties(),
```
indention

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const unsigned int number_of_control_points = GetGeometry().size();
+
+        if (rResult.size() != 3 * number_of_control_points)
+            rResult.resize(3 * number_of_control_points, false);
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        // Resizing as needed the LHS
+        unsigned int mat_size = number_of_control_points * 3;
+
+        if (rDampingMatrix.size1() != mat_size)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+
+        // 1.-Get Damping Coeffitients (RAYLEIGH_ALPHA, RAYLEIGH_BETA)
+        double alpha = 0.0;
+        if (GetProperties().Has(RAYLEIGH_ALPHA))
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        double alpha = 0.0;
+        if (GetProperties().Has(RAYLEIGH_ALPHA))
+            alpha = GetProperties()[RAYLEIGH_ALPHA];
+        else if (rCurrentProcessInfo.Has(RAYLEIGH_ALPHA))
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+            alpha = rCurrentProcessInfo[RAYLEIGH_ALPHA];
+
+        double beta = 0.0;
+        if (GetProperties().Has(RAYLEIGH_BETA))
```
brackets

philbucher replied:
for one line ifs we usually don't use brackets

philbucher replied:
or use the [ternary operator](https://en.wikipedia.org/wiki/%3F:#C++):
`double beta = GetProperties().Has(RAYLEIGH_BETA) ? GetProperties()[RAYLEIGH_BETA] : 0.0;`

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        double beta = 0.0;
+        if (GetProperties().Has(RAYLEIGH_BETA))
+            beta = GetProperties()[RAYLEIGH_BETA];
+        else if (rCurrentProcessInfo.Has(RAYLEIGH_BETA))
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rOutput.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rOutput.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rOutput.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rOutput.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rOutput.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rVariable == CONSTITUTIVE_LAW)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        if (rValues.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+    {
+        const unsigned int size = GetGeometry().IntegrationPoints().size();
+
+        if (rValues.size() != 1)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const unsigned int number_of_control_points = GetGeometry().size();
+        const unsigned int mat_size = number_of_control_points * 3;
+
+        if (rValues.size() != mat_size)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        const unsigned int number_of_control_points = GetGeometry().size();
+        const unsigned int mat_size = number_of_control_points * 3;
+
+        if (rValues.size() != mat_size)
```
brackets

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+        KRATOS_TRY
+
+        //Constitutive Law initialisation
+        if (mConstitutiveLawVector.size() != 1)
```
I recommend to use brackets also for one-liner

```
} else {
    // ...
}
```

philbucher replied:
we usually don't do that

loumalouomega replied:
We do in fact @roigcarlo

philbucher replied:
is this documented in the StyleGuide?
Would be nice then we don't have to discuss it again each time...

oberbichler commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,559 @@
+    {
+        KRATOS_TRY;
+
+        const unsigned int number_of_control_points = GetGeometry().size();
```
I would avoid the usage of unsigned datatypes since they have no benefits and are highly error prone

Personally I would also drop the `const` if it is clear for the compiler and the user

philbucher replied:
for me it is often not clear whether sth is const or not
I usually but it where possible, since it can prevent errors

tteschemacher committed:
Small design changes

Identation modified
unsigned int removed

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,503 @@
+    {
+        KRATOS_TRY;
+
+        const int number_of_control_points = GetGeometry().size();
```
usually we use `std::size_t` (wrapper as `SizeType`) for this

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,503 @@
+
+        const unsigned int pos = this->GetGeometry()[0].GetDofPosition(DISPLACEMENT_X);
+
+        for (unsigned int i = 0; i < number_of_control_points; ++i) {
```
usually we use `std::size_t` (wrapper as `IndexType`) for this

philbucher commented on applications/IGAStructuralMechanicsApplication/custom_elements/base_discrete_element.cpp:
```diff
@@ -0,0 +1,503 @@
+            rOutput.resize(1);
+
+        if (rVariable == VELOCITY) {
+            const int& number_of_control_points = GetGeometry().size();
```
I would remove the reference here if you don't explicitly need it.
First it is faster afaik (int takes 32 bit while a reference takes 64 bit) but more important it tells whoever reads this code that sth "special" here is going on that require the use of a reference

philbucher reviewed (commented):
some comments

oberbichler commented:
We should wait with this pull request until the `IgaStructuralMechanicsApplication` is updated to `pybind`.

philbucher commented:
can this be closed?

philbucher commented:
is this branch still needed?
